### PR TITLE
[READY] Space Dragon Quality Improvement Update

### PR DESCRIPTION
**Pull request by IndieanaJones**
## About The Pull Request

This PR updates space dragon will a combination of code improvements, QOL improvements, bugfixes, and balance updates.  This ended being a lot more than I intended to do, so let's get started.

**BUGFIXES**
- Fixed a bug where Space Dragon dying didn't destroy the rifts due to a runtime error
- Fixed an oversight which allowed Space Dragon to constantly tear at the same wall, allowing it to spam noise constantly
- Fixed a bug which allowed a killed Space Dragon to summon rifts again upon revival
- Fixed a bug where certain interactions with stamina attacks would remove Space Dragon's speed boost while enraged
- Fixed a bug where Space Dragon wouldn't be slowed from failing its objective if the reason it failed was death

**QOL**
- Space Dragon plays a noise when vanishing due to not placing a portal in time
- Gust is no longer tied to an ability button, instead used via right click
- Some text has been replaced to be less awkward

**CODE IMPROVEMENTS**
- Refactored Space Dragon's speed modifiers are now unique modifiers instead of just changing varspeeds.  This fixes several issues and also allows SD's mob speed to be edited in the future.
- Added some variables to adjust some features of SD's gust
- Added a variable to adjust carp rift carp spawn interval

**BALANCE CHANGES**
- Time it takes to charge a rift decreased from 8 minutes to 5 minutes
- Carp spawns now occur every 30 seconds from every 40 seconds
- Space Dragon now receives permanent rage upon completing its objective
- Space Dragon's fire breath is now 33% faster to travel
- Space Dragon winning no longer calls the shuttle
- Space Dragon now automatically completes its objective should the shuttle arrive while its still alive.  It also gains all the benefits that it would normally from completing it (perma speed, infinite carp from rifts)
- Gust now reaches 4 tiles away, increased from 3

**SPRITES/SOUNDS**
- Added new custom death roar
- Removed now unused rift ability icon

## Why It's Good For The Game

All these changes intend to improve the space dragon experience via more consistency and making everything feel better to play.  The biggest balance change is definitely making Space Dragon win if the shuttle arrives while its active, forcing the crew to think carefully about how they handle the situation as opposed to outright avoiding it.

## Changelog
:cl:
qol: Space Dragon's gust attack is now tied to right click instead of an ability button
balance: Space Dragon has received numerous balance changes, most notably that Space Dragon will stall the shuttle until it loses or completes its objectives
fix: Fixed numerous bugs with Space Dragon, most notably the rifts not vanishing after Space Dragon's death
soundadd: Space Dragon now has a custom roar for dying and winning
imagedel: Deleted now unused Space Dragon ability icon
refactor: Space Dragon's speed changes are now controlled using movespeed modifiers instead of direct speed adjustments
/:cl:

### CODE REVIEW

**Comment by bluetshirtguy** [comment]
Did you really have to make it cause a hostile environment? That's such a slog man.

**Comment by MrMelbert** [comment]
I don't like the idea of forcing hostile environment with something that's timegated. If blob is uncontested they can rush to their victory but if a dragon is uncontested you literally have to delay the round by 10 minutes while they sit there and charge their portals. That's just boring.

**Comment by tralezab** [comment]
Atomize this man, this has fixes, refactors, new sounds, and balance/rebalance in the same pr

**Comment by Serxule** [comment]
i like all of this except for shuttle locking, remove that and its good

**Comment by IndieanaJones** [comment]
OK, alternate idea: having the shuttle arrive while space dragon is active uncaps the spawn rate on the portals and gives the dragon perma-rage as if he won, while also completing his objective.  I don't like the fact people can simply choose to ignore the fact SD is active, so as long as we can get that solved I'm fine.

**Comment by IndieanaJones** [comment]
Change: Space Dragon no longer blocks the shuttle from leaving.  However, if the shuttle arrives while Space Dragon is still alive, Space Dragon automatically completes its objective and gets all the benefits from doing so normally.

**Comment by bluetshirtguy** [comment]
> Change: Space Dragon no longer blocks the shuttle from leaving. However, if the shuttle arrives while Space Dragon is still alive, Space Dragon automatically completes its objective and gets all the benefits from doing so normally.

Why are you so intent on the crew feeling forced or being forced to deal with him? That kinda thinking you have hampers the kind of freedom of choice and sandbox play players enjoy.

**Comment by IndieanaJones** [comment]
> Why are you so intent on the crew feeling forced or being forced to deal with him? That kinda thinking you have hampers the kind of freedom of choice and sandbox play players enjoy.

Why force the crew to fight the blob?  Why does a cult winning end the round?  Why not give players the ability to disable ever being the assassination target of a traitor?  The point is that this is a game which uses conflicts to drive engagement and intrigue, and this is clearly such a case.  If the crew calls a shuttle on a blob or a cult, they both punish that behavior because it attempts to circumvent their existence entirely, and while this is a sandbox game, we can ask the players to engage with the mechanics on some level; to make everything optional would be absurd, as then nothing would have meaning, creating a hollow shell of an experience.

The current issue with space dragon is the fact that he is intrinsically tied to the rifts he creates.  If a crew decides to call shuttle because SD exists, its no different than calling shuttle on a blob.  Space Dragon is designed so that the rift essentially acts as a leash for it; it can't murderbone across the station because if it does, then the rift is destroyed and it immediately gets a nerf to its combat capabilities, making it an easy target for anyone looking to take it down.  However, because of this and the time it takes for a space dragon to charge its rifts, you can call a shuttle, have it arrive, leave on it, and be at Centcomm before the rifts finish charging.  If you're thinking optimal and don't mind ending the round early, the most foolproof way to deal with Space Dragon is not to deal with it at all; there's no consequence to simply ignoring it outside of some carp attacking you, but they're some of the weakest simplemobs in the game and they too require time to actually spawn in from the rifts.

So, it becomes evidently clear that some reason to motivate engagement between dragon and crew has become necessary, as the dragon is an loud-style antagonist that works best when the crew is tasked with directly engaging it with conflict.  Causing it not to call the shuttle would have worked for this situation as the crew wouldn't be able to leave otherwise, essentially making it unavoidable.  However, I understand the point made that if there is no sufficient force to challenge the dragon, then waiting for what could be up to 15 minutes for the dragon to slowly complete his objective could be boring, whereas something like blob increases its ability exponentially to end the round if nobody steps up to oppose it.  So, the solution is simple, give the players incentive to fight the dragon or destroy the rift before the shuttle comes, but still let them have it as an option (but deinsentivized).

With this solution now implemented, the crew makes their decision based on not overall winning or losing the round, but something almost identical in value; their lives.  If they attempt to use the shuttle to circumvent SD now, they need to have a plan for when it inevitably attacks the shuttle with greater carp assistance than normal.  They can answer this question any number of ways, all of which require engagement with the game's mechanics and different jobs may offer different solutions to it, which creates engagement with the game's mechanics and an enjoyable scenario.  Alternatively, it might incentivize players to seek the obvious aggressive solution of killing the dragon or destroying the rift, which the game offers a plethora of solutions to do so.  The way you paint what I'm doing here as burdensome would only make sense to the most cowardly of players, as essentially you don't want to do anything and still not worry about a threat placed in front of you.  A traitor who has to kill you certainly won't intend on sparing you if they want to complete their objectives, and the space dragon is instinctively put at odds with the crew from design alone, so I doubt it'll spare you either.

So, yeah, I'm totally down with encouraging the crew to feel forced to come up with a solution to deal with the dragon.  Conflicts like these are backbone to SS13 as a core and problem-solving is something we come to expect from players when it comes to dealing with antagonists.  The fact Space Dragon didn't do this before was arguably bad design which is now being rectified, so if you don't want to "deal with him" then you better find a good hiding spot on the shuttle or an escape pod.

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 1]
```diff
@@ -434,46 +456,31 @@
...
 	button_icon_state = "carp_rift"
-	desc = "Summon a rift to bring forth a horde of space carp."
 
-/datum/action/innate/space_dragon/summon_rift/Activate()
+/datum/action/innate/summon_rift/Activate()
 	var/mob/living/simple_animal/hostile/space_dragon/S = owner
 	if(S.using_special)
 		return
+	if(S.riftTimer == -1)
+		to_chat(S, "<span class='warning'>Your death has left you unable to summon rifts!</span>")
```
Comment (Ghommie) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:483]:
```suggestion
		to_chat(S, "<span class='warning'>Your death has left you unable to summon rifts!</span>")
		return
```

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 2]
```diff
@@ -597,40 +604,40 @@
...
 		notify_ghosts("The carp rift can summon an additional carp!", source = src, action = NOTIFY_ORBIT, flashwindow = FALSE, header = "Carp Spawn Available")
-		last_carp_inc -= 40
+		last_carp_inc -= carp_interval
 
 	// Is the rift now fully charged?
 	if(time_charged >= max_charge)
 		charge_state = CHARGE_COMPLETED
 		var/area/A = get_area(src)
-		priority_announce("Spatial object has reached peak energy charge in [initial(A.name)], please stand-by.", "Central Command Spatial Corps")
+		priority_announce("Spatial object has reached peak energy charge in [initial(A.name)], please stand-by.", "Central Command Wildlife Observations")
```
Comment (Ghommie) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:620]: The announcement subject is still a "Spatial object" despite the title being about wildlife.

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 3]
```diff
@@ -434,46 +456,31 @@
...
+	rift_empower(TRUE, TRUE)
 	var/datum/antagonist/space_dragon/S = mind.has_antag_datum(/datum/antagonist/space_dragon)
 	if(S)
 		var/datum/objective/summon_carp/main_objective = locate() in S.objectives
 		if(main_objective)
 			main_objective.completed = TRUE
-	sound_to_playing_players('sound/machines/alarm.ogg')
-	sleep(100)
-	priority_announce("A large amount of lifeforms have been detected approaching [station_name()] at extreme speeds. Evacuation of the remaining crew will begin immediately.", "Central Command Spatial Corps")
+	priority_announce("A large amount of lifeforms have been detected approaching [station_name()] at extreme speeds. Remaining crew are advised to evacuate as soon as possible.", "Central Command Wildlife Observations")
```
Comment (Ghommie) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:465]: It niggles me a little, but about every other priority announcement about space pests around or within the station uses the "Lifesign Alert" title. I'm not asking you to change it, but I believe announcement titles will need some recoherence in the future in the form of defines.

**Comment by SuperNovaa41** [comment]
It doesn't really seem like you're balancing anything, it's just a pure space dragon buff

**Comment by IndieanaJones** [comment]
> It doesn't really seem like you're balancing anything, it's just a pure space dragon buff

Yeah

**Comment by nianjiilical** [comment]
dragons biggest issue is how much the crew is incentivized to just call the shuttle and not engage with it so i fully support all of these changes

if possible itd be nice to add some text so people know that the dragon greentexts/rages if the shuttle arrives, though idk a good non intrusive/logical way to do it

**Comment by Serxule** [comment]
> Why force the crew to fight the blob? Why does a cult winning end the round? Why not give players the ability to disable ever being the assassination target of a traitor? The point is that this is a game which uses conflicts to drive engagement and intrigue, and this is clearly such a case. If the crew calls a shuttle on a blob or a cult, they both punish that behavior because it attempts to circumvent their existence entirely, and while this is a sandbox game, we can ask the players to engage with the mechanics on some level; to make everything optional would be absurd, as then nothing would have meaning, creating a hollow shell of an experience.
> 
I get what you mean but still, dragons are relatively easily killed, one person can kill a dragon if they're smart enough, its not nearly a big enough threat for centcomm themselves to lock the shuttle, the reason it does with cult and blob is both of those could take over centcomm while a dragon cannot

**Comment by IndieanaJones** [comment]
> I get what you mean but still, dragons are relatively easily killed, one person can kill a dragon if they're smart enough, its not nearly a big enough threat for centcomm themselves to lock the shuttle, the reason it does with cult and blob is both of those could take over centcomm while a dragon cannot

Well that's not an issue now.

**Comment by Fikou** [comment]
smh putting refactor to get +2 points for a balance pr

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 4]
```diff
@@ -16,7 +16,7 @@
...
  * Space Dragon can summon only one rift at a time, and can do so anywhere a blob is allowed to spawn.  In order to trigger his victory condition, Space Dragon must summon and defend three rifts while they charge.
  * Space Dragon, when spawned, has five minutes to summon the first rift.  Failing to do so will cause Space Dragon to return from whence he came.
- * When the rift spawns, ghosts can interact with it to spawn in as space carp to help complete the mission.  One carp is granted when the rift is first summoned, with an extra one every 40 seconds.
+ * When the rift spawns, ghosts can interact with it to spawn in as space carp to help complete the mission.  One carp is granted when the rift is first summoned, with an extra one every 30 seconds.
  * Once the victory condition is met, the shuttle is called and all current rifts are allowed to spawn infinite sentient space carp.
  * If a charging rift is destroyed, Space Dragon will be incredibly slowed, and the endlag on his gust attack is greatly increased on each use.
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:21]: add the shuttle mechanics to this doc?

**Comment by carshalash** [comment]
Buffing your own antag even more so than before and then adding in punishment for trying to flee from it is kinda cringe.

**Comment by Sparkezel** [comment]
Dunno how about you but I feel like all antags could use a buff right now

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm (hunk 4 above)
Comment (IndieanaJones) [reply code/modules/mob/living/simple_animal/hostile/space_dragon.dm:21]: Yeah, thanks for the reminder <3

**Comment by IndieanaJones** [comment]
> Buffing your own antag even more so than before and then adding in punishment for trying to flee from it is kinda cringe.

I don't know if this has occurred to you, but there isn't a lot of interest in balancing out the midround antagonists.  I could essentially claim they're all mine at this point because nobody else is touching them.  Also, shaming someone for buffing something they created is the dumbest shit because what do you want me to do, wait for someone else to balance it?  Nah, as its creator it should be my responsibility to see that balanced is achieved, not push it onto someone else because some numbnut (you in this case) will perceive it as self-serving bias.

For reference, the buffs included here were made on the basis of feedback I had received about the dragon's current performance.  The fire "buff" is a double-edged sword because it means that the fire also dissipates faster now as well, creating more opening between fire breaths.  Gust being a bit stronger isn't going to kill anyone either, it was made mostly so it would require people to dodge it intentionally a little bit more now, but realistically an additional tile isn't going to do all that much. 
 Therefore, these two buffs are so minor I really doubt it'll impact the combat effectiveness in most situations, but having the fire be more responsive at the cost of not lingering as long seems like a great change for both crew and dragon.  Finally, rift charging was way too slow beforehand, 8 minutes in hindsight is extremely ridiculous and felt pretty bad on the SD end, so I think almost everyone could agree that its a good change.

The punishment for trying to ignore Space Dragon, as I argued in my last diatribe, is a much better step from not having any penalty in terms of design.  The dragon appearing before simply held no weight because it'd be established in one spot and the crew could very easily leave it be and get back to Centcomm before the dragon finished.  With this change, it establishes an interest in the dragon's existence for everyone because now, like almost all of our other antagonists, letting it live can and will impact you negatively.  I understand adding more threat to an antagonist scares some of our playerbase, but its simply better this way.  You'll need one hell of an argument to convince more otherwise.

Also, just because its a big scary dragon doesn't mean its insanely strong.  While you can kill the dragon, you can also just destroy the rift, and the game offers a shitton of solutions for both ways so your option aren't limited.  Blow up the dragon, blow up the rift, have someone distract the dragon while you destroy the portal, its not as bad as you probably think it is.  Ideally this could get a test merge so I can gauge the strength of the changes for balancing before merging, so if you're afraid the dragon will become some god-tier antag from these changes then fear not, because we have tools like that to make sure things like this don't completely imbalance the game.

**Comment by optimumtact** [comment]
We shoudl delete them all, they are cringe and crap

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 5]
```diff
@@ -360,31 +375,39 @@
  * Empowers and depowers Space Dragon after a successful rift charge.
  * Empowered, Space Dragon regains all his health and becomes temporarily faster for 30 seconds, along with being tinted red.
  * Depowered simply resets him back to his default state.
+ * Arguments:
+ * * is_empowered - Whether or not we're adding the buff or removing the buff
+ * * is_permanent - Only applies if is_empowered is true, whether or not the buff should be removed after a time.
  */
-/mob/living/simple_animal/hostile/space_dragon/proc/rift_empower(is_empowered)
+/mob/living/simple_animal/hostile/space_dragon/proc/rift_empower(is_empowered, is_permanent)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:382]: Can you make this a bitfield instead?

[I don't particularly like boolean parameters anyway](https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3), but if we're going to have them they should at least be readable.

Alternatively, splitting up the bools into separate procs is more prefered if you're willing to put in more work, especially `is_empowered` which looks to alter the *entire* behavior.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Minor code changes, I'll trust you know what you're doing with the balance changes. Having dragons be more of a priority for the crew with the shuttle is a good thing, I didn't think there was anything wrong with having it block the shuttle from leaving honestly.

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 6]
```diff
@@ -550,11 +574,9 @@
...
 	STOP_PROCESSING(SSobj, src)
 	if(time_charged != max_charge + 1)
-		to_chat(dragon, "<span class='boldwarning'>The rift has been destroyed!  You have failed, and find yourself brought down by the weight of your failure.</span>")
-		dragon.set_varspeed(5)
-		dragon.tiredness_mult = 5
-		dragon.destroy_rifts()
-		playsound(src, 'sound/vehicles/rocketlaunch.ogg', 100, TRUE)
+		dragon?.destroy_rifts()
+		if(dragon)
+			to_chat(dragon, "<span class='boldwarning'>A rift has been destroyed!  You have failed, and find yourself weakened.</span>")
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/hostile/space_dragon.dm:579]:
```suggestion
			to_chat(dragon, "<span class='boldwarning'>A rift has been destroyed! You have failed, and find yourself weakened.</span>")
```

The `if` is also unnecessary, you can `to_chat` null and it's just a no-op, but doesn't matter.

**Comment by IndieanaJones** [comment]
The integration error seems to be referring to an issue with mannitol.

**Comment by Mothblocks** [comment]
It's a runtime that there's a PR to fix, it's not related to this.

It passed before my text edit, I'll take a gamble.

**Comment by carshalash** [comment]
Very epic to make the death noise literally just ear rape.

**Comment by IndieanaJones** [comment]
> Very epic to make the death noise literally just ear rape.

A genuine mistake on my part, thank goodness I had this page open to see you comment this here since you decided bitching about it on Discord without at least @ing me about it was the best way to see it fixed.

**Comment by optimumtact** [comment]
all carsh does is subtweet people about how he has issues

**Pull request merged**